Introduce leaf protocols to prevent leaf nodes from being casted (#2108)
apple/swift-syntax: Release Notes/510.md

## API Behavior Changes

- Leaf Node Protocols

>   - Description: Syntax nodes that do not act as base nodes for other syntax types now conform to a new protocols that deprecates inherited casting methods (e.g. `AccessorBlockSyntax` to `LeafSyntaxNodeProtocol` or `AccessorDeclSyntax` to `LeafDeclSyntaxNodeProtocol`). This change aims to prevent unsafe type-casting by issuing deprecation warnings for methods that will result in failed casts.

Reviewer: I would try to avoid the introduction of the protocols since it’s just a technical detail.

```suggestion
  - Description: Syntax nodes that do not act as base nodes for other syntax types have the casting methods marked as deprecated. This prevents unsafe type-casting by issuing deprecation warnings for methods that will always result in failed casts.
```

Author: Thanks, done

## API Behavior Changes

- Leaf Node Casts
  - Description: Syntax nodes that do not act as base nodes for other syntax types have the casting methods marked as deprecated. This prevents unsafe type-casting by issuing deprecation warnings for methods that will always result in failed casts.
  - Issue: https://github.com/apple/swift-syntax/issues/2092
  - Pull Request: https://github.com/apple/swift-syntax/pull/2108

- Same-Type Casts 
  - Description: `is`, `as`, and `cast` overloads on `SyntaxProtocol` with same-type conversions are marked as deprecated. The deprecated methods emit a warning indicating the cast will always succeed.
  - Issue: https://github.com/apple/swift-syntax/issues/2092
  - Pull Request: https://github.com/apple/swift-syntax/pull/2108

- Base Node Casts
  - Description: `is`, `as`, and `cast` methods on base node protocols with base-type conversions are marked as deprecated. The deprecated methods emit a warning that informs the developer that the cast will always succeed and should be done using the base node's initializer.
  - Issue: https://github.com/apple/swift-syntax/issues/2092
  - Pull Request: https://github.com/apple/swift-syntax/pull/2108

> 

Reviewer: I think this should go in the `Deprecations` sections since the runtime behavior of the API doesn’t change.

Author: Done
